lib/segment/src/data_types/query_context.rs

```diff
@@ -83,7 +84,7 @@ impl QueryContext {
         SegmentQueryContext {
             query_context: self,
             deleted_points: None,
-            hardware_counter: HardwareCounterCell::default(),
+            hardware_counter: Rc::new(HardwareCounterCell::new()),
```

Comment:
Why RC is now needed?

Reply:
To not clone `SegmentQueryContext` and accidentally create multiple instances of `hardware_counter`, of which we consume one and drop the others causing a debug panic or (wrong) warning.
It also can't be a simple reference, because [here](https://github.com/qdrant/qdrant/blob/7e39692d80acfbf10b527c8664d2d4e79667578e/lib/segment/src/data_types/query_context.rs#L87) we need to create a `SegmentQueryContext` with a new instance of `HardwareCounterCell`.

If you have another suggestion, please let me know!

Reply:
Maybe forbid clone?

Reply:
We currently use to clone it [here](https://github.com/qdrant/qdrant/blob/master/lib/collection/src/collection_manager/holders/proxy_segment.rs#L355) and might need it in future in similar cases too, especially as `SegmentQueryContext` only holds borrowed values (except HardwareCounterCell).
Is there a specific reason you don't think RC is a good fit here?

Reply:
I mean, forbid clone of the `HardwareCounterCell` and use reference

Reply:
We can't use a reference because then we won't be able to create a new `SegmentQueryContext` [here](https://github.com/qdrant/qdrant/blob/7e39692d80acfbf10b527c8664d2d4e79667578e/lib/segment/src/data_types/query_context.rs#L87) as it would return a reference to the HardwareCounterCell which is only a temporary value

Reply:
`SegmentQueryContext` exists strictly inside a single thread, while QueryContext is shared.
If we want to get results from multiple segments, it should be aggregated

Reply:
In that case I think it is not avoidable to use `RC` here, unless we make `SegmentQueryContext` non-clone and find a workaround for the existing code calling `clone`.
I prefer the RC approach because it gives us more flexibility without real drawbacks.
What do you think?

Reply:
> In that case I think it is not avoidable to use RC here, unless we make SegmentQueryContext non-clone and find a workaround for the existing code calling clone.

something like this:
https://github.com/qdrant/qdrant/pull/5288/commits/e269f68cb567193ecba6df0cc8ed37988716739b